pinot-core/src/main/java/org/apache/pinot/core/data/manager/offline/DimensionTableDataManager.java

```diff
@@ -0,0 +1,168 @@
+   * instance should be properly initialized via {@link #init} method before using.
+   */
+  public static DimensionTableDataManager createInstanceByTableName(String tableNameWithType) {
+    return _instances.computeIfAbsent(tableNameWithType, k -> new DimensionTableDataManager());
+  }
+
+  public static DimensionTableDataManager getInstanceByTableName(String tableNameWithType) {
+    return _instances.get(tableNameWithType);
+  }
+
+  /**
+   * Instance properties/methods
+   */
+  private final ReadWriteLock _rwl = new ReentrantReadWriteLock();
+  private final Lock _lookupTableReadLock = _rwl.readLock();
+  private final Lock _lookupTableWriteLock = _rwl.writeLock();
+
+  // _lookupTable is a HashMap used for storing/serving records for a table keyed by table PK
+  @GuardedBy("_rwl")
+  private final Map<PrimaryKey, GenericRow> _lookupTable = new HashMap<>();
+  private Schema _tableSchema;
+  private List<String> _primaryKeyColumns;
+
+  @Override
+  protected void doInit() {
+    super.doInit();
+
+    _tableSchema = ZKMetadataProvider.getTableSchema(_propertyStore, _tableNameWithType);
+    Preconditions.checkState(_tableSchema != null, "Failed to find schema for table: %s", _tableNameWithType);
+    _primaryKeyColumns = _tableSchema.getPrimaryKeyColumns();
+    Preconditions.checkState(!CollectionUtils.isEmpty(_primaryKeyColumns),
+        "Primary key columns must be configured for dimension table: %s", _tableNameWithType);
+  }
+
+  @Override
+  public void addSegment(File indexDir, IndexLoadingConfig indexLoadingConfig)
+      throws Exception {
+    super.addSegment(indexDir, indexLoadingConfig);
+    try {
+      loadLookupTable();
+      _logger.info("Successfully added segment {} and loaded lookup table: {}", indexDir.getName(), getTableName());
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Error loading lookup table: %s", getTableName()), e);
+    }
+  }
+
+  @Override
+  public void removeSegment(String segmentName) {
+    super.removeSegment(segmentName);
+    try {
+      loadLookupTable();
+      _logger.info("Successfully removed segment {} and reloaded lookup table: {}", segmentName, getTableName());
+    } catch (Exception e) {
+      throw new RuntimeException(String
+          .format("Error reloading lookup table after segment remove ({}) for table: {}", segmentName, getTableName()),
+          e);
+    }
+  }
+
+  /**
+   * `loadLookupTable()` reads contents of the DimensionTable into _lookupTable HashMap for fast lookup.
+   */
+  private void loadLookupTable()
+      throws Exception {
+    _lookupTableWriteLock.lock();
+    try {
+      _lookupTable.clear();
+      List<SegmentDataManager> segmentManagers = acquireAllSegments();
+      if (segmentManagers.size() == 0) {
+        return;
+      }
+
+      try {
+        for (SegmentDataManager segmentManager : segmentManagers) {
```

Comment:
Since we want that a lookup table should be small, it may be worth asserting that the table has to be in one segment?

Reply:
Right, I listed the table size issue in the description as one of the things I'd like to addresses subsequently.

What I'm thinking is putting a 'default maximum' value for Dimension table size quota, say something like 200MB. But I also think this value should be configurable via server config (by the Pinot admin, not the user). This should allow use cases for larger tables if you have the hardware resources to handle them. I think restricting the table to be a single segment is somewhat arbitrary and doesn't help us that much. Happy to hear you and @Jackie-Jiang 's thoughts though.

Reply:
OK yes, we can address it later.